templates/.gitlab-ci.yml

```diff
@@ -0,0 +1,171 @@
+# SPDX-License-Identifier: LGPL-3.0+
+#
+# Copyright (C) 2018 Ultimaker B.V.
+# Copyright (C) 2018 Raymond Siudak <r.siudak@ultimaker.com>
+#
+# GitLab CI package build is split up into two parts.
+# * Build environment: creates the docker image that is needed for package generation
+# * Package generation: creates the final artifact
+
+stages:
+  - lint
+  - prepare
+  - push
+  - build
+  - test
+  - cleanup
+
+# Common requisites
+# =================
+.parameters_common: &parameters_common
+  tags:
+    - docker
+
+.changes_docker: &changes_docker
+  changes:
+    - Dockerfile
+    - .dockerignore
+    - dockerfiles/**/*
+
+.shell_linting_common: &shell_linting_common
+  <<: *parameters_common
+  image: registry.hub.docker.com/koalaman/shellcheck-alpine:stable
+  stage: lint
+
+lint_shell_scripts:
+  <<: *shell_linting_common
+  script:
+    - shellcheck -C -f tty -s sh "build"*".sh"
+    - find 'scripts/' -iname '*.sh' -exec shellcheck -x -C -f tty -s sh {} \;
+
+# Build environment
+# The 'prepare' and the 'push' stages are executed conditionally. They both
+# include 'build_environment_common' that makes sure that the stage is only
+# performed when docker changes are performed on any branch.
+# =================
+.build_environment_common: &environment_common
+  <<: *parameters_common
+  image: registry.hub.docker.com/library/docker:stable
+
+prepare_build_environment:
+  <<: *environment_common
+  only:
+    <<: *changes_docker
+  stage: prepare
+  script:
+    - docker build --rm -t "${CI_COMMIT_SHA}:${CI_PIPELINE_ID}" .
+    - |
+        docker run \
+        --rm \
+        --privileged \
+        -e "ARM_EMU_BIN=${ARM_EMU_BIN}" \
+        -v "${ARM_EMU_BIN}:${ARM_EMU_BIN}:ro" \
```

Comment:
Seems kind of specific to programs compiled with ARM?

Reply:
no, but this does add the arm emulator to be _available_; of course, this section can be removed in a more generic non-embedded setting.

Reply:
btw, with the proper flag on the host, this is not even needed actually. The 'fixup' flag.
```console
cat /proc/sys/fs/binfmt_misc/qemu-arm | grep flags
flags: OCF
```